### DISCUSSION TITLE
creating custom meshes on runtime

### DISCUSSION THREAD

**Discussion by DelinWorks** [axmolengine/axmol#680] [Q&A] [answered]
Hi! Can someone help me with this peace of code?

In this code i'm trying to render a custom runtime created mesh, firstly I set the vertex attributes and what not and then I fill the vertices array and then the index array, I then add that generated mesh to the sprite3d and a material program state and a texture to the sprite3d, I've tried everything but that sprite3d doesn't seem to be rendered at all no matter what I do, Any help would be appreciated 🙂

![image](https://user-images.githubusercontent.com/45469625/176557245-e1311179-05aa-4fa9-a1b6-a619081a3bc6.png)

A black or textured triangle should show in the middle of the green and red lines, I even tried disabling depth testing and still nothing shows.

```
sp3d = Sprite3D::create();

int perVertexSizeInFloat = 8;  // 3+3+2

std::vector<float> vertices = {
    // Front Triangle 
    // Position         Normal          UV
       0,100,0,         1,1,1,          0,1,
       100,0,0,         1,1,1,          1,0,
       0,0,0,           1,1,1,          0,1,

    // Back Triangle
    // Position         Normal          UV
       0,0,0,           1,1,1,          0,1,
       100,0,0,         1,1,1,          1,0,
       0,100,0,         1,1,1,          0,1,
};

MeshData::IndexArray indices = { 0, 1, 2 };

std::vector<MeshVertexAttrib> attribs;
MeshVertexAttrib att;

att.type = backend::VertexFormat::FLOAT3;
att.vertexAttrib = shaderinfos::VertexKey::VERTEX_ATTRIB_POSITION;
attribs.push_back(att);

att.type = backend::VertexFormat::FLOAT3;
att.vertexAttrib = shaderinfos::VertexKey::VERTEX_ATTRIB_NORMAL;
attribs.push_back(att);

att.type = backend::VertexFormat::FLOAT2;
att.vertexAttrib = shaderinfos::VertexKey::VERTEX_ATTRIB_TEX_COORD;
attribs.push_back(att);

auto mat = Sprite3DMaterial::createBuiltInMaterial(Sprite3DMaterial::MaterialType::UNLIT, false);

auto mesh = Mesh::create(vertices, perVertexSizeInFloat, indices, attribs);
sp3d->addMesh(mesh);

sp3d->setMaterial(mat);
sp3d->setTexture("Untitled.png");

worldNode->addChild(sp3d, 10);

sp3d->setPosition3D(Vec3(0, 0, 0));
```

that's all called in `init()` function of the scene

What I discovered is that after creating a `MeshData` object and then passing vertex data and attribs to it and then I create a `MeshVertexData` using that `MeshData` the mesh index at 0 using `getMeshIndexDataByIndex(0)` doesn't seem to have any vertex data in it like in this picture

![image](https://user-images.githubusercontent.com/45469625/176515722-c2355ab6-0ea1-4615-a1a6-1412daf73272.png)

Maybe the vertex data has been uploaded to the GPU Memory and deleted from CPU Memory? I can't tell :(

I'm starting to suspect that the mesh creation process is faulty.

**Comment by DelinWorks** [comment, marked as answer]
Nevermind turns out it was a layering problem that `setGlobalZOrder()` didn't fix I had to force 2D queue for it to work.

**Comment by aismann** [comment]
@DelinWorks ,
maybe a nice demo on cpp-test?
Here it will be forgotten I belive.
